cmd/dogstatsd/main.go

```diff
@@ -88,26 +88,28 @@ func init() {
 }
 
 func start(cmd *cobra.Command, args []string) error {
+	// Main context passed to components
+	ctx, cancel := context.WithCancel(context.Background())
+
+	defer func(cancel context.CancelFunc) {
+		stopAgent(cancel)
+	}(cancel)
+
+	stopCh := make(chan struct{})
+	go handleSignals(stopCh)
 
-	mainCtx, mainCtxCancel, err := runAgent()
+	err := runAgent(ctx)
 	if err != nil {
 		return err
 	}
-	// Setup a channel to catch OS signals
-	signalCh := make(chan os.Signal, 1)
-	signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM)
 
-	// Block here until we receive the interrupt signal
-	<-signalCh
+	// Block here until we receive a stop signal
+	<-stopCh
```

Comment:
Instead of blocking on `stopCh`, isn't calling `handleSignals()` without `stopCh` enough to block until a signal is received?

Reply:
The process exits as soon as the main goroutine finishes executing, so we need to wait here (otherwise the process would exit immediately after it has finished its startup sequence).

Reply:
What I suggested was to replace the lines https://github.com/DataDog/datadog-agent/blob/1ffcd3c8a7a776903981892b48f1cadf47302e02/cmd/dogstatsd/main.go#L98-L107 by
```
	err := runAgent(ctx)
	if err != nil {
		return err
	}

	handleSignals()
```
with
```
func handleSignals() { // No argument
	// Setup a channel to catch OS signals
	signalCh := make(chan os.Signal, 1)
	signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM, syscall.SIGPIPE)

	// Block here until we receive the interrupt signal
	for signo := range signalCh {
		switch signo {
		case syscall.SIGPIPE:
			// By default systemd redirects the stdout to journald. When journald is stopped or crashes we receive a SIGPIPE signal.
			// Go ignores SIGPIPE signals unless it is when stdout or stdout is closed, in this case the agent is stopped.
			// We never want dogstatsd to stop upon receiving SIGPIPE, so we intercept the SIGPIPE signals and just discard them.
		default:
			log.Infof("Received signal '%s', shutting down...", signo)
			return
		}
	}
}
```
```

Reply:
I prefer setting up signal handling before initiating the startup sequence (`runAgent`) since so that `SIGPIPE` signals sent before dogstatsd has finished starting up don't stop dogstatsd. (the startup sequence takes a few seconds to complete)

Reply:
Make sense. Thanks for the explanation!